In [1]:
from Stitchr import stitchrfunctions as fxn
from Stitchr import stitchr as st
import pandas as pd
import wandb
import pandas as pd
from dotenv import load_dotenv
import os
import subprocess
import numpy as np

This is the official documentatin if one would like to use Stitchr in an individual pipeline: 
[Stitchr Doc](https://github.com/JamieHeather/stitchr/blob/main/docs/importing.rst)

In [2]:
chain = "TRB"
species = "HUMAN"

tcr_dat, functionality, partial = fxn.get_imgt_data(chain, st.gene_types, species)
codons = fxn.get_optimal_codons('', species)

V_segment = "TRBV9*01"
J_segment = "TRBJ2-2*01"
CDR3_sequence_beta = "CASSENTANTGELFF"

tcr_bits = {"v": V_segment, "j": J_segment, "cdr3": CDR3_sequence_beta,
            "l": "TRBV7-3*01", "c": "TRBC1*01",
            "skip_c_checks": False, "species": species, "seamless": False,
            "5_prime_seq": "", "3_prime_seq": "", "name": "TCR"}

stitched = st.stitch(tcr_bits, tcr_dat, functionality, partial, codons, 3, "")
stitched

(['TCR',
  'TRBV9*01',
  'TRBJ2-2*01',
  'TRBC1*01',
  'CASSENTANTGELFF',
  'TRBV7-3*01(L)'],
 'ATGGGCACCAGGCTCCTCTGCTGGGCAGCCCTGTGCCTCCTGGGGGCAGATCACACAGATTCTGGAGTCACACAAACCCCAAAGCACCTGATCACAGCAACTGGACAGCGAGTGACGCTGAGATGCTCCCCTAGGTCTGGAGACCTCTCTGTGTACTGGTACCAACAGAGCCTGGACCAGGGCCTCCAGTTCCTCATTCAGTATTATAATGGAGAAGAGAGAGCAAAAGGAAACATTCTTGAACGATTCTCCGCACAACAGTTCCCTGACTTGCACTCTGAACTAAACCTGAGCTCTCTGGAGCTGGGGGACTCAGCTTTGTATTTCTGTGCCAGCAGCGAGAACACCGCCAACACCGGGGAGCTGTTTTTTGGAGAAGGCTCTAGGCTGACCGTACTGGAGGACCTGAACAAGGTGTTCCCACCCGAGGTCGCTGTGTTTGAGCCATCAGAAGCAGAGATCTCCCACACCCAAAAGGCCACACTGGTGTGCCTGGCCACAGGCTTCTTCCCCGACCACGTGGAGCTGAGCTGGTGGGTGAATGGGAAGGAGGTGCACAGTGGGGTCAGCACGGACCCGCAGCCCCTCAAGGAGCAGCCCGCCCTCAATGACTCCAGATACTGCCTGAGCAGCCGCCTGAGGGTCTCGGCCACCTTCTGGCAGAACCCCCGCAACCACTTCCGCTGTCAAGTCCAGTTCTACGGGCTCTCGGAGAATGACGAGTGGACCCAGGATAGGGCCAAACCCGTCACCCAGATCGTCAGCGCCGAGGCCTGGGGTAGAGCAGACTGTGGCTTTACCTCGGTGTCCTACCAGCAAGGGGTCCTGTCTGCCACCATCCTCTATGAGATCCTGCTAGGGAAGGCCACCCTGTATGCTGTGCTGGTCAGCGCCCTTGTGTTGA

However the documentation mentions that for high throughput (as we must use) one should use [*thimble*](https://github.com/JamieHeather/stitchr/blob/main/docs/thimble.rst)

In the follwoing we try to convert a whole subset of the VDJdb using this approach:

In [3]:
load_dotenv()
PROJECT_NAME = os.getenv("MAIN_PROJECT_NAME")
MCPAS_TCR = os.getenv("MCPAS_TCR_ARTIFACT")
VDJDB_BETA = os.getenv("VDJDB_BETA_ARTIFACT")
VDJDB_PAIRED = os.getenv("VDJDB_PAIRED_ARTIFACT")
wandb.init(project=PROJECT_NAME, job_type="data-Stitchr-pre-processing")

Failed to detect the name of this notebook, you can set it manually with the WANDB_NOTEBOOK_NAME environment variable to enable code saving.
wandb: Currently logged in as: eggervali (ba-zhaw). Use `wandb login --relogin` to force relogin


Here we first create only beta-chain,only alpha-chain or only paired-chain datasets and store them in a .tsv!

This first try is by using only paired data! Then we transform the dataset according to the documentation so that it fits the [template for the documentation](https://github.com/JamieHeather/stitchr/tree/main/templates)

In [4]:
vdjdb_paired_artifact = wandb.use_artifact(f"{VDJDB_PAIRED}:latest")
vdjdb_paired_table = vdjdb_paired_artifact.get(f"{VDJDB_PAIRED}_table.table.json")
vdjdb_paired_df = pd.DataFrame(data=vdjdb_paired_table.data, columns=vdjdb_paired_table.columns)
vdjdb_paired_df

wandb: Downloading large artifact VDJdb_paired_only:latest, 54.56MB. 1 files... 
wandb:   1 of 1 files downloaded.  
Done. 0:0:0.3


complex.id Gene                  CDR3            V           J  \
0              1  TRB  CASSYLPGQGDHYSNQPQHF    TRBV13*01  TRBJ1-5*01   
1              2  TRB   CASSFEPGQGFYSNQPQHF    TRBV13*01  TRBJ1-5*01   
2              3  TRB  CASSYEPGQVSHYSNQPQHF    TRBV13*01  TRBJ1-5*01   
3              4  TRB        CASSALASLNEQFF    TRBV14*01  TRBJ2-1*01   
4              5  TRB  CASSYLPGQGDHYSNQPQHF    TRBV13*01  TRBJ1-5*01   
...          ...  ...                   ...          ...         ...   
56233      30590  TRA         CMDEGGSNYKLTF  TRAV26-1*01   TRAJ53*01   
56234      30591  TRA           CSLYNNNDMRF  TRAV26-1*01   TRAJ43*01   
56235      30592  TRA         CALSTDSWGKLQF     TRAV6*01   TRAJ24*01   
56236      30593  TRA          CAPQGATNKLIF  TRAV12-2*01   TRAJ32*01   
56237      30594  TRA        CLVGGSGGYNKLIF     TRAV4*01    TRAJ4*01   

           Species           MHC A           MHC B MHC class        Epitope  \
0      HomoSapiens        HLA-B*08             B2M      MHCI       FLKEKGGL   
1      HomoSapiens        HLA-B*08             B2M      MHCI       FLKEKGGL   
2      HomoSapiens        HLA-B*08             B2M      MHCI       FLKEKGGL   
3      HomoSapiens        HLA-B*08             B2M      MHCI       FLKEKGGL   
4      HomoSapiens        HLA-B*08             B2M      MHCI       FLKEQGGL   
...            ...             ...             ...       ...            ...   
56233  HomoSapiens  HLA-DQA1*05:01  HLA-DQB1*02:01     MHCII   PQPELPYPQPQL   
56234  HomoSapiens  HLA-DQA1*05:01  HLA-DQB1*02:01     MHCII   PQPELPYPQPQL   
56235  HomoSapiens  HLA-DQA1*05:01  HLA-DQB1*02:01     MHCII  PQQPFPQPEQPFP   
56236  HomoSapiens  HLA-DQA1*05:01  HLA-DQB1*02:01     MHCII  PQQPFPQPEQPFP   
56237  HomoSapiens  HLA-DQA1*05:01  HLA-DQB1*02:01     MHCII  PQQPFPQPEQPFP   

      Epitope gene Epitope species      Reference  \
0              Nef           HIV-1  PMID:15596521   
1              Nef           HIV-1  PMID:15596521   
2              Nef           HIV-1  PMID:15596521   
3              Nef           HIV-1  PMID:15596521   
4              Nef           HIV-1  PMID:15596521   
...            ...             ...            ...   
56233       Gluten           Wheat  PMID:33927715   
56234       Gluten           Wheat  PMID:33927715   
56235       Gluten           Wheat  PMID:33927715   
56236       Gluten           Wheat  PMID:33927715   
56237       Gluten           Wheat  PMID:33927715   

                                                  Method  \
0      {"frequency": "", "identification": "tetramer-...   
1      {"frequency": "", "identification": "tetramer-...   
2      {"frequency": "", "identification": "tetramer-...   
3      {"frequency": "", "identification": "tetramer-...   
4      {"frequency": "", "identification": "tetramer-...   
...                                                  ...   
56233  {"frequency": "", "identification": "tetramer-...   
56234  {"frequency": "", "identification": "tetramer-...   
56235  {"frequency": "", "identification": "tetramer-...   
56236  {"frequency": "", "identification": "tetramer-...   
56237  {"frequency": "", "identification": "tetramer-...   

                                                    Meta  \
0      {"cell.subset": "CD8+", "clone.id": "", "donor...   
1      {"cell.subset": "CD8+", "clone.id": "", "donor...   
2      {"cell.subset": "CD8+", "clone.id": "", "donor...   
3      {"cell.subset": "CD8+", "clone.id": "", "donor...   
4      {"cell.subset": "CD8+", "clone.id": "", "donor...   
...                                                  ...   
56233  {"cell.subset": "CD4+", "clone.id": "", "donor...   
56234  {"cell.subset": "CD4+", "clone.id": "", "donor...   
56235  {"cell.subset": "CD4+", "clone.id": "", "donor...   
56236  {"cell.subset": "CD4+", "clone.id": "", "donor...   
56237  {"cell.subset": "CD4+", "clone.id": "", "donor...   

                                                 CDR3fix Score  
0      {"cdr3": "CASSYLPGQGDHYSNQPQHF", "cdr3_ol

In [5]:
# Splitting the DataFrame into TRB and TRA based on the line transition
line_transition = 28120  # Line where information about beta chains termiante
vdjdb_TRB_df = vdjdb_paired_df.iloc[:line_transition-1].copy()  # TRB chains information
vdjdb_TRA_df = vdjdb_paired_df.iloc[line_transition-1:].copy()  # TRA chains information

# Renaming columns for TRB DataFrame
vdjdb_TRB_df.rename(columns={
    "complex.id": "TCR_name", 
    "V": "TRBV", 
    "J": 'TRBJ', 
    'CDR3': 'TRB_CDR3'
}, inplace=True)

vdjdb_TRA_df.rename(columns={
    'complex.id': 'TCR_name', 
    'V': 'TRAV', 
    'J': 'TRAJ', 
    'CDR3': 'TRA_CDR3'
}, inplace=True)

In [6]:
vdjdb_TRB_df.columns

Index(['TCR_name', 'Gene', 'TRB_CDR3', 'TRBV', 'TRBJ', 'Species', 'MHC A',
       'MHC B', 'MHC class', 'Epitope', 'Epitope gene', 'Epitope species',
       'Reference', 'Method', 'Meta', 'CDR3fix', 'Score'],
      dtype='object')

In [7]:
vdjdb_TRA_df.columns

Index(['TCR_name', 'Gene', 'TRA_CDR3', 'TRAV', 'TRAJ', 'Species', 'MHC A',
       'MHC B', 'MHC class', 'Epitope', 'Epitope gene', 'Epitope species',
       'Reference', 'Method', 'Meta', 'CDR3fix', 'Score'],
      dtype='object')

In [8]:
vdjdb_paired_one_line_df = pd.merge(vdjdb_TRB_df, vdjdb_TRA_df, on=['TCR_name', "Species", "MHC A", "MHC B", "MHC class", "Epitope", "Epitope gene", "Epitope species"])
vdjdb_paired_one_line_df.rename(columns={
    'Gene_x': 'Gene_TRB',
    'Gene_y': 'Gene_TRA',
    'Reference_x': 'Reference_TRB',
    'Method_x': 'Method_TRB',
    'Meta_x': 'Meta_TRB',
    'CDR3fix_x': 'CDR3fix_TRB',
    'Score_x': 'Score_TRB',
    'Reference_y': 'Reference_TRA',
    'Method_y': 'Method_TRA',
    'Meta_y': 'Meta_TRA',
    'CDR3fix_y': 'CDR3fix_TRA',
    'Score_y': 'Score_TRA',
}, inplace=True)
# complete dataframe with all columns (Stitchr + native from VDJdb) => However, not needed so no save neccessary until now!
# vdjdb_paired_one_line_df.to_csv("../data/VDJdb/temp/vdjdb_paired_one_line.tsv", sep="\t", index=False)
vdjdb_paired_one_line_df

TCR_name Gene_TRB              TRB_CDR3         TRBV        TRBJ  \
0            1      TRB  CASSYLPGQGDHYSNQPQHF    TRBV13*01  TRBJ1-5*01   
1            2      TRB   CASSFEPGQGFYSNQPQHF    TRBV13*01  TRBJ1-5*01   
2            3      TRB  CASSYEPGQVSHYSNQPQHF    TRBV13*01  TRBJ1-5*01   
3            4      TRB        CASSALASLNEQFF    TRBV14*01  TRBJ2-1*01   
4            5      TRB  CASSYLPGQGDHYSNQPQHF    TRBV13*01  TRBJ1-5*01   
...        ...      ...                   ...          ...         ...   
28114    30590      TRB         CASSVRSTDTQYF   TRBV7-2*01  TRBJ2-3*01   
28115    30591      TRB         CASSLRYTDTQYF   TRBV7-2*01  TRBJ2-3*01   
28116    30592      TRB       CASSPGQGGDNEQFF   TRBV7-3*01  TRBJ2-1*01   
28117    30593      TRB       CASSLGAGGQETQYF   TRBV5-1*01  TRBJ2-5*01   
28118    30594      TRB         CASSSTAQETQYF  TRBV11-2*01  TRBJ2-5*01   

           Species           MHC A           MHC B MHC class        Epitope  \
0      HomoSapiens        HLA-B*08             B2M      MHCI       FLKEKGGL   
1      HomoSapiens        HLA-B*08             B2M      MHCI       FLKEKGGL   
2      HomoSapiens        HLA-B*08             B2M      MHCI       FLKEKGGL   
3      HomoSapiens        HLA-B*08             B2M      MHCI       FLKEKGGL   
4      HomoSapiens        HLA-B*08             B2M      MHCI       FLKEQGGL   
...            ...             ...             ...       ...            ...   
28114  HomoSapiens  HLA-DQA1*05:01  HLA-DQB1*02:01     MHCII   PQPELPYPQPQL   
28115  HomoSapiens  HLA-DQA1*05:01  HLA-DQB1*02:01     MHCII   PQPELPYPQPQL   
28116  HomoSapiens  HLA-DQA1*05:01  HLA-DQB1*02:01     MHCII  PQQPFPQPEQPFP   
28117  HomoSapiens  HLA-DQA1*05:01  HLA-DQB1*02:01     MHCII  PQQPFPQPEQPFP   
28118  HomoSapiens  HLA-DQA1*05:01  HLA-DQB1*02:01     MHCII  PQQPFPQPEQPFP   

       ... Score_TRB Gene_TRA        TRA_CDR3             TRAV       TRAJ  \
0      ...         2      TRA   CIVRAPGRADMRF      TRAV26-1*01  TRAJ43*01   
1      ...         2      TRA  CAVPSGAGSYQLTF        TRAV20*01  TRAJ28*01   
2      ...         2      TRA     CAVKASGSRLT         TRAV2*01        nan   
3      ...         2      TRA   CAYRPPGTYKYIF  TRAV38-2/DV8*01  TRAJ40*01   
4      ...         2      TRA   CIVRAPGRADMRF      TRAV26-1*01  TRAJ43*01   
...    ...       ...      ...             ...              ...        ...   
28114  ...         0      TRA   CMDEGGSNYKLTF      TRAV26-1*01  TRAJ53*01   
28115  ...         0      TRA     CSLYNNNDMRF      TRAV26-1*01  TRAJ43*01   
28116  ...         0      TRA   CALSTDSWGKLQF         TRAV6*01  TRAJ24*01   
28117  ...         2      TRA    CAPQGATNKLIF      TRAV12-2*01  TRAJ32*01   
28118  ...         0      TRA  CLVGGSGGYNKLIF         TRAV4*01   TRAJ4*01   

       Reference_TRA                                         Method_TRA  \
0      PMID:15596521  {"frequency": "", "identification": "tetramer-...   
1      PMID:15596521  {"frequency": "", "identification": "tetramer-...   
2      PMID:15596521  {"frequency": "", "identification": "tetramer-...   
3      PMID:15596521  {"frequency": "", "identification": "tetramer-...   
4      PMID:15596521  {"frequency": "", "identification": "tetramer-...   
...              ...                                                ...   
28114  PMID:33927715  {"frequency": "", "identification": "tetramer-...   
28115  PMID:33927715  {"frequency": "", "identification": "tetramer-...   
28116  PMID:33927715  {"frequency": "", "identification": "tetramer-...   
28117  PMID:33927715  {"frequency": "", "identification": "tetramer-...   
28118  PMID:33927715  {"frequency": "", "identification": "tetramer-...   

                                                Meta_TRA  \
0      {"cell.subset": "CD8+", "clone.id": "", "donor...   
1      {"cell.subset": "CD8+", "clone.id": "", "donor...   
2      {"cell.subset": "CD8+", "clone.id": "", "donor...   
3      {"cell.subset": "CD8+", "clone.id": "", "donor...   
4      {"cell.subset": "CD8+", "clone.id":

In [9]:
vdjdb_paired_one_line_df['TRA_leader'] = vdjdb_paired_one_line_df['TRAV'] + "(L)"
vdjdb_paired_one_line_df['TRB_leader'] = vdjdb_paired_one_line_df['TRBV'] + "(L)"

for new_column in ['TRAC', 'TRBC', 'Linker', 'Link_order', 'TRA_5_prime_seq', 'TRA_3_prime_seq', 'TRB_5_prime_seq', 'TRB_3_prime_seq']:
    vdjdb_paired_one_line_df[new_column] = np.nan # TODO check this

template_columns_order = [
    'TCR_name', 'TRAV', 'TRAJ', 'TRA_CDR3', 'TRBV', 'TRBJ', 'TRB_CDR3',
    'TRAC', 'TRBC', 'TRA_leader', 'TRB_leader', 'Linker', 'Link_order',
    'TRA_5_prime_seq', 'TRA_3_prime_seq', 'TRB_5_prime_seq', 'TRB_3_prime_seq'
]

vdjdb_paired_only_Stitchr_schema_df = vdjdb_paired_one_line_df[template_columns_order]

vdjdb_paired_only_Stitchr_schema_df.to_csv("../data/VDJdb/temp/vdjdb_paired_only_Stitchr_schema.tsv", sep="\t", index=False)

vdjdb_paired_only_Stitchr_schema_df

TCR_name             TRAV       TRAJ        TRA_CDR3         TRBV  \
0            1      TRAV26-1*01  TRAJ43*01   CIVRAPGRADMRF    TRBV13*01   
1            2        TRAV20*01  TRAJ28*01  CAVPSGAGSYQLTF    TRBV13*01   
2            3         TRAV2*01        nan     CAVKASGSRLT    TRBV13*01   
3            4  TRAV38-2/DV8*01  TRAJ40*01   CAYRPPGTYKYIF    TRBV14*01   
4            5      TRAV26-1*01  TRAJ43*01   CIVRAPGRADMRF    TRBV13*01   
...        ...              ...        ...             ...          ...   
28114    30590      TRAV26-1*01  TRAJ53*01   CMDEGGSNYKLTF   TRBV7-2*01   
28115    30591      TRAV26-1*01  TRAJ43*01     CSLYNNNDMRF   TRBV7-2*01   
28116    30592         TRAV6*01  TRAJ24*01   CALSTDSWGKLQF   TRBV7-3*01   
28117    30593      TRAV12-2*01  TRAJ32*01    CAPQGATNKLIF   TRBV5-1*01   
28118    30594         TRAV4*01   TRAJ4*01  CLVGGSGGYNKLIF  TRBV11-2*01   

             TRBJ              TRB_CDR3  TRAC  TRBC          TRA_leader  \
0      TRBJ1-5*01  CASSYLPGQGDHYSNQPQHF   NaN   NaN      TRAV26-1*01(L)   
1      TRBJ1-5*01   CASSFEPGQGFYSNQPQHF   NaN   NaN        TRAV20*01(L)   
2      TRBJ1-5*01  CASSYEPGQVSHYSNQPQHF   NaN   NaN         TRAV2*01(L)   
3      TRBJ2-1*01        CASSALASLNEQFF   NaN   NaN  TRAV38-2/DV8*01(L)   
4      TRBJ1-5*01  CASSYLPGQGDHYSNQPQHF   NaN   NaN      TRAV26-1*01(L)   
...           ...                   ...   ...   ...                 ...   
28114  TRBJ2-3*01         CASSVRSTDTQYF   NaN   NaN      TRAV26-1*01(L)   
28115  TRBJ2-3*01         CASSLRYTDTQYF   NaN   NaN      TRAV26-1*01(L)   
28116  TRBJ2-1*01       CASSPGQGGDNEQFF   NaN   NaN         TRAV6*01(L)   
28117  TRBJ2-5*01       CASSLGAGGQETQYF   NaN   NaN      TRAV12-2*01(L)   
28118  TRBJ2-5*01         CASSSTAQETQYF   NaN   NaN         TRAV4*01(L)   

           TRB_leader  Linker  Link_order  TRA_5_prime_seq  TRA_3_prime_seq  \
0        TRBV13*01(L)     NaN         NaN              NaN              NaN   
1        TRBV13*01(L)     NaN         NaN              NaN              NaN   
2        TRBV13*01(L)     NaN         NaN              NaN              NaN   
3        TRBV14*01(L)     NaN         NaN              NaN              NaN   
4        TRBV13*01(L)     NaN         NaN              NaN              NaN   
...               ...     ...         ...              ...              ...   
28114   TRBV7-2*01(L)     NaN         NaN              NaN              NaN   
28115   TRBV7-2*01(L)     NaN         NaN              NaN              NaN   
28116   TRBV7-3*01(L)     NaN         NaN              NaN              NaN   
28117   TRBV5-1*01(L)     NaN         NaN              NaN              NaN   
28118  TRBV11-2*01(L)     NaN         NaN              NaN              NaN   

       TRB_5_prime_seq  TRB_3_prime_seq  
0                  NaN              NaN  
1                  NaN              NaN  
2                  NaN              NaN  
3                  NaN              NaN  
4                  NaN              NaN  
...                ...              ...  
28114              NaN              NaN  
28115              NaN              NaN  
28116              NaN              NaN  
28117              NaN              NaN  
28118              NaN              NaN  

[28119 rows x 17 columns]

In [10]:
vdjdb_paired_only_Stitchr_schema_df.columns

Index(['TCR_name', 'TRAV', 'TRAJ', 'TRA_CDR3', 'TRBV', 'TRBJ', 'TRB_CDR3',
       'TRAC', 'TRBC', 'TRA_leader', 'TRB_leader', 'Linker', 'Link_order',
       'TRA_5_prime_seq', 'TRA_3_prime_seq', 'TRB_5_prime_seq',
       'TRB_3_prime_seq'],
      dtype='object')

Below is the desired schema, according to the documentation! We need to prepare the df to fit into this schema.

In [11]:
template_stitchr = pd.read_csv("../data/StitchrOutputs/templates/thimble_input_example_TRA-TRB.tsv", sep="\t")
template_stitchr.columns

Index(['TCR_name', 'TRAV', 'TRAJ', 'TRA_CDR3', 'TRBV', 'TRBJ', 'TRB_CDR3',
       'TRAC', 'TRBC', 'TRA_leader', 'TRB_leader', 'Linker', 'Link_order',
       'TRA_5_prime_seq', 'TRA_3_prime_seq', 'TRB_5_prime_seq',
       'TRB_3_prime_seq'],
      dtype='object')

In [12]:
str(template_stitchr.columns) == str(vdjdb_paired_only_Stitchr_schema_df.columns)

True

This command needs to be adjusted if not paired-chain data gets passed!

- AB -> paired (alpha & beta chains)
- B -> ONLY beta chains
- A -> ONLY alpha chains

In [13]:
from_paired_vdjdb_path = "../data/VDJdb/temp/vdjdb_paired_only_Stitchr_schema.tsv"
to_paired_vdjdb_path =  "../data/StitchrOutputs/VDJdb_Stitchr/VDJdb_paired_Stitchr"
command = ['thimble', '-in', from_paired_vdjdb_path, '-r', 'AB', "-s", "HUMAN", "-o", to_paired_vdjdb_path]

In [14]:
result = subprocess.run(command, capture_output=True, text=True)

# Check if the command was executed successfully
if result.returncode == 0:
    print("Command executed successfully.")
    # If you want to print the output
    print("Output:", result.stdout)
else:
    print("Error in command execution.")
    # If you want to print the error
    print("Error:", result.stderr)

Command executed successfully.
Output: Took 25.39 seconds



In [15]:
df_vdjdb_stitchr = pd.read_csv(to_paired_vdjdb_path+".tsv", sep="\t")
df_vdjdb_stitchr

TCR_name                                             TRA_nt  \
0             1  ATGAGGCTGGTGGCAAGAGTAACTGTGTTTCTGACCTTTGGAACTA...   
1             2  ATGGAGAAAATGTTGGAGTGTGCATTCATAGTCTTGTGGCTTCAGC...   
2             3                                                NaN   
3             4  ATGGCATGCCCTGGCTTCCTGTGGGCACTTGTGATCTCCACCTGTC...   
4             5  ATGAGGCTGGTGGCAAGAGTAACTGTGTTTCTGACCTTTGGAACTA...   
...         ...                                                ...   
28114     30590  ATGAGGCTGGTGGCAAGAGTAACTGTGTTTCTGACCTTTGGAACTA...   
28115     30591  ATGAGGCTGGTGGCAAGAGTAACTGTGTTTCTGACCTTTGGAACTA...   
28116     30592  ATGGAGTCATTCCTGGGAGGTGTTTTGCTGATTTTGTGGCTTCAAG...   
28117     30593  ATGAAATCCTTGAGAGTTTTACTAGTGATCCTGTGGCTTCAGTTGA...   
28118     30594  ATGAGGCAAGTGGCGAGAGTGATCGTGTTCCTGACCCTGAGTACTT...   

                                                  TRB_nt  \
0      ATGCTTAGTCCTGACCTGCCTGACTCTGCCTGGAACACCAGGCTCC...   
1      ATGCTTAGTCCTGACCTGCCTGACTCTGCCTGGAACACCAGGCTCC...   
2      ATGCTTAGTCCTGACCTGCCTGACTCTGCCTGGAACACCAGGCTCC...   
3      ATGGTTTCCAGGCTTCTCAGTTTAGTGTCCCTTTGTCTCCTGGGAG...   
4      ATGCTTAGTCCTGACCTGCCTGACTCTGCCTGGAACACCAGGCTCC...   
...                                                  ...   
28114  ATGGGCACCAGGCTCCTCTTCTGGGTGGCCTTCTGTCTCCTGGGGG...   
28115  ATGGGCACCAGGCTCCTCTTCTGGGTGGCCTTCTGTCTCCTGGGGG...   
28116  ATGGGCACCAGGCTCCTCTGCTGGGCAGCCCTGTGCCTCCTGGGGG...   
28117  ATGGGCTCCAGGCTGCTCTGTTGGGTGCTGCTTTGTCTCCTGGGAG...   
28118  ATGGGCACCAGGCTCCTCTGCTGGGCGGCCCTCTGTCTCCTGGGAG...   

                                                  TRA_aa  \
0      MRLVARVTVFLTFGTIIDAKTTQPPSMDCAEGRAANLPCNHSTISG...   
1      MEKMLECAFIVLWLQLGWLSGEDQVTQSPEALRLQEGESSSLNCSY...   
2                                                    NaN   
3      MACPGFLWALVISTCLEFSMAQTVTQSQPEMSVQEAETVTLSCTYD...   
4      MRLVARVTVFLTFGTIIDAKTTQPPSMDCAEGRAANLPCNHSTISG...   
...                                                  ...   
28114  MRLVARVTVFLTFGTIIDAKTTQPPSMDCAEGRAANLPCNHSTISG...   
28115  MRLVARVTVFLTFGTIIDAKTTQPPSMDCAEGRAANLPCNHSTISG...   
28116  MESFLGGVLLILWLQVDWVKSQKIEQNSEALNIQEGKTATLTCNYT...   
28117  MKSLRVLLVILWLQLSWVWSQQKEVEQNSGPLSVPEGAIASLNCTY...   
28118  MRQVARVIVFLTLSTLSLAKTTQPISMDSYEGQEVNITCSHNNIAT...   

                                                  TRB_aa             TRAV  \
0      MLSPDLPDSAWNTRLLCHVMLCLLGAVSVAAGVIQSPRHLIKEKRE...      TRAV26-1*01   
1      MLSPDLPDSAWNTRLLCHVMLCLLGAVSVAAGVIQSPRHLIKEKRE...        TRAV20*01   
2      MLSPDLPDSAWNTRLLCHVMLCLLGAVSVAAGVIQSPRHLIKEKRE...         TRAV2*01   
3      MVSRLLSLVSLCLLGAKHIEAGVTQFPSHSVIEKGQTVTLRCDPIS...  TRAV38-2/DV8*01   
4      MLSPDLPDSAWNTRLLCHVMLCLLGAVSVAAGVIQSPRHLIKEKRE...      TRAV26-1*01   
...                                                  ...              ...   
28114  MGTRLLFWVAFCLLGADHTGAGVSQSPSNKVTEKGKDVELRCDPIS...      TRAV26-1*01   
28115  MGTRLLFWVAFCLLGADHTGAGVSQSPSNKVTEKGKDVELRCDPIS...      TRAV26-1*01   
28116  MGTRLLCWAALCLLGADHTGAGVSQTPSNKVTEKGKYVELRCDPIS...         TRAV6*01   
28117  MGSRLLCWVLLCLLGAGPVKAGVTQTPRYLIKTRGQQVTLSCSPIS...      TRAV12-2*01   
28118  MGTRLLCWAALCLLGAELTEAGVAQSPRYKIIEKRQSVAFWCNPIS...         TRAV4*01   

            TRAJ        TRA_CDR3         TRBV        TRBJ  ...  \
0      TRAJ43*01   CIVRAPGRADMRF    TRBV13*01  TRBJ1-5*01  ...   
1      TRAJ28*01  CAVPSGAGSYQLTF    TRBV13*01  TRBJ1-5*01  ...   
2            NaN     CAVKASGSRLT    TRBV13*01  TRBJ1-5*01  ...   
3      TRAJ40*01   CAYRPPGTYKYIF    TRBV14*01  TRBJ2-1*01  ...   
4      TRAJ43*01   CIVRAPGRADMRF    TRBV13*01  TRBJ1-5*01  ...   
...          ...             ...          ...         ...  ...   
28114  TRAJ53*01   CMDEGGSNYKLTF   TRBV7-2*01  TRBJ2-3*01  ...   
28115  TRAJ43*01     CSLYNNNDMRF   TRBV7-2*01  TRBJ2-3*01  ...   
28116  TRAJ24*01   CALSTDSWGKLQF   TRBV7-3*01  TRBJ2-1*01  ...   
28117  TRAJ32*01    CAPQGATNKLIF   TRBV5-1*01  TRBJ2-5*01  ...   
28118   TRAJ4*01  CLVGGSGGYNKLIF  TRBV11-2*01  TRBJ2-5*01  ...   

      

-------------------------------------------------------------------------------------------------
After doing the paired dataset, we want to to the same pre-processing step (Stitchr) with the beta-only!

In [16]:
vdjdb_beta_artifact = wandb.use_artifact(f"{VDJDB_BETA}:latest")
vdjdb_beta_table = vdjdb_beta_artifact.get(f"{VDJDB_BETA}_table.table.json")
vdjdb_beta_df = pd.DataFrame(data=vdjdb_beta_table.data, columns=vdjdb_beta_table.columns)
vdjdb_beta_df

wandb:   1 of 1 files downloaded.  


complex.id Gene                  CDR3            V           J  \
0              1  TRB  CASSYLPGQGDHYSNQPQHF    TRBV13*01  TRBJ1-5*01   
1              0  TRB   CASSFEAGQGFFSNQPQHF    TRBV13*01  TRBJ1-5*01   
2              2  TRB   CASSFEPGQGFYSNQPQHF    TRBV13*01  TRBJ1-5*01   
3              3  TRB  CASSYEPGQVSHYSNQPQHF    TRBV13*01  TRBJ1-5*01   
4              4  TRB        CASSALASLNEQFF    TRBV14*01  TRBJ2-1*01   
...          ...  ...                   ...          ...         ...   
49037      30590  TRB         CASSVRSTDTQYF   TRBV7-2*01  TRBJ2-3*01   
49038      30591  TRB         CASSLRYTDTQYF   TRBV7-2*01  TRBJ2-3*01   
49039      30592  TRB       CASSPGQGGDNEQFF   TRBV7-3*01  TRBJ2-1*01   
49040      30593  TRB       CASSLGAGGQETQYF   TRBV5-1*01  TRBJ2-5*01   
49041      30594  TRB         CASSSTAQETQYF  TRBV11-2*01  TRBJ2-5*01   

           Species           MHC A           MHC B MHC class        Epitope  \
0      HomoSapiens        HLA-B*08             B2M      MHCI       FLKEKGGL   
1      HomoSapiens        HLA-B*08             B2M      MHCI       FLKEKGGL   
2      HomoSapiens        HLA-B*08             B2M      MHCI       FLKEKGGL   
3      HomoSapiens        HLA-B*08             B2M      MHCI       FLKEKGGL   
4      HomoSapiens        HLA-B*08             B2M      MHCI       FLKEKGGL   
...            ...             ...             ...       ...            ...   
49037  HomoSapiens  HLA-DQA1*05:01  HLA-DQB1*02:01     MHCII   PQPELPYPQPQL   
49038  HomoSapiens  HLA-DQA1*05:01  HLA-DQB1*02:01     MHCII   PQPELPYPQPQL   
49039  HomoSapiens  HLA-DQA1*05:01  HLA-DQB1*02:01     MHCII  PQQPFPQPEQPFP   
49040  HomoSapiens  HLA-DQA1*05:01  HLA-DQB1*02:01     MHCII  PQQPFPQPEQPFP   
49041  HomoSapiens  HLA-DQA1*05:01  HLA-DQB1*02:01     MHCII  PQQPFPQPEQPFP   

      Epitope gene Epitope species      Reference  \
0              Nef           HIV-1  PMID:15596521   
1              Nef           HIV-1  PMID:15596521   
2              Nef           HIV-1  PMID:15596521   
3              Nef           HIV-1  PMID:15596521   
4              Nef           HIV-1  PMID:15596521   
...            ...             ...            ...   
49037       Gluten           Wheat  PMID:33927715   
49038       Gluten           Wheat  PMID:33927715   
49039       Gluten           Wheat  PMID:33927715   
49040       Gluten           Wheat  PMID:33927715   
49041       Gluten           Wheat  PMID:33927715   

                                                  Method  \
0      {"frequency": "", "identification": "tetramer-...   
1      {"frequency": "", "identification": "tetramer-...   
2      {"frequency": "", "identification": "tetramer-...   
3      {"frequency": "", "identification": "tetramer-...   
4      {"frequency": "", "identification": "tetramer-...   
...                                                  ...   
49037  {"frequency": "", "identification": "tetramer-...   
49038  {"frequency": "", "identification": "tetramer-...   
49039  {"frequency": "", "identification": "tetramer-...   
49040  {"frequency": "", "identification": "tetramer-...   
49041  {"frequency": "", "identification": "tetramer-...   

                                                    Meta  \
0      {"cell.subset": "CD8+", "clone.id": "", "donor...   
1      {"cell.subset": "CD8+", "clone.id": "", "donor...   
2      {"cell.subset": "CD8+", "clone.id": "", "donor...   
3      {"cell.subset": "CD8+", "clone.id": "", "donor...   
4      {"cell.subset": "CD8+", "clone.id": "", "donor...   
...                                                  ...   
49037  {"cell.subset": "CD4+", "clone.id": "", "donor...   
49038  {"cell.subset": "CD4+", "clone.id": "", "donor...   
49039  {"cell.subset": "CD4+", "clone.id": "", "donor...   
49040  {"cell.subset": "CD4+", "clone.id": "", "donor...   
49041  {"cell.subset": "CD4+", "clone.id": "", "donor...   

                                                 CDR3fix Score  
0      {"cdr3": "CASSYLPGQGDHYSNQPQHF", "cdr3_ol

In [17]:
vdjdb_beta_only_df = vdjdb_beta_df

In [18]:
vdjdb_beta_only_df.rename(columns={
    "complex.id": "TCR_name", 
    "V": "TRBV", 
    "J": 'TRBJ', 
    'CDR3': 'TRB_CDR3'
}, inplace=True)
vdjdb_beta_only_df.columns

Index(['TCR_name', 'Gene', 'TRB_CDR3', 'TRBV', 'TRBJ', 'Species', 'MHC A',
       'MHC B', 'MHC class', 'Epitope', 'Epitope gene', 'Epitope species',
       'Reference', 'Method', 'Meta', 'CDR3fix', 'Score'],
      dtype='object')

In [19]:
template_stitchr.columns

Index(['TCR_name', 'TRAV', 'TRAJ', 'TRA_CDR3', 'TRBV', 'TRBJ', 'TRB_CDR3',
       'TRAC', 'TRBC', 'TRA_leader', 'TRB_leader', 'Linker', 'Link_order',
       'TRA_5_prime_seq', 'TRA_3_prime_seq', 'TRB_5_prime_seq',
       'TRB_3_prime_seq'],
      dtype='object')

In [20]:
vdjdb_beta_only_df['TRB_leader'] = vdjdb_beta_only_df['TRBV'] + "(L)"
vdjdb_beta_only_df

TCR_name Gene              TRB_CDR3         TRBV        TRBJ  \
0            1  TRB  CASSYLPGQGDHYSNQPQHF    TRBV13*01  TRBJ1-5*01   
1            0  TRB   CASSFEAGQGFFSNQPQHF    TRBV13*01  TRBJ1-5*01   
2            2  TRB   CASSFEPGQGFYSNQPQHF    TRBV13*01  TRBJ1-5*01   
3            3  TRB  CASSYEPGQVSHYSNQPQHF    TRBV13*01  TRBJ1-5*01   
4            4  TRB        CASSALASLNEQFF    TRBV14*01  TRBJ2-1*01   
...        ...  ...                   ...          ...         ...   
49037    30590  TRB         CASSVRSTDTQYF   TRBV7-2*01  TRBJ2-3*01   
49038    30591  TRB         CASSLRYTDTQYF   TRBV7-2*01  TRBJ2-3*01   
49039    30592  TRB       CASSPGQGGDNEQFF   TRBV7-3*01  TRBJ2-1*01   
49040    30593  TRB       CASSLGAGGQETQYF   TRBV5-1*01  TRBJ2-5*01   
49041    30594  TRB         CASSSTAQETQYF  TRBV11-2*01  TRBJ2-5*01   

           Species           MHC A           MHC B MHC class        Epitope  \
0      HomoSapiens        HLA-B*08             B2M      MHCI       FLKEKGGL   
1      HomoSapiens        HLA-B*08             B2M      MHCI       FLKEKGGL   
2      HomoSapiens        HLA-B*08             B2M      MHCI       FLKEKGGL   
3      HomoSapiens        HLA-B*08             B2M      MHCI       FLKEKGGL   
4      HomoSapiens        HLA-B*08             B2M      MHCI       FLKEKGGL   
...            ...             ...             ...       ...            ...   
49037  HomoSapiens  HLA-DQA1*05:01  HLA-DQB1*02:01     MHCII   PQPELPYPQPQL   
49038  HomoSapiens  HLA-DQA1*05:01  HLA-DQB1*02:01     MHCII   PQPELPYPQPQL   
49039  HomoSapiens  HLA-DQA1*05:01  HLA-DQB1*02:01     MHCII  PQQPFPQPEQPFP   
49040  HomoSapiens  HLA-DQA1*05:01  HLA-DQB1*02:01     MHCII  PQQPFPQPEQPFP   
49041  HomoSapiens  HLA-DQA1*05:01  HLA-DQB1*02:01     MHCII  PQQPFPQPEQPFP   

      Epitope gene Epitope species      Reference  \
0              Nef           HIV-1  PMID:15596521   
1              Nef           HIV-1  PMID:15596521   
2              Nef           HIV-1  PMID:15596521   
3              Nef           HIV-1  PMID:15596521   
4              Nef           HIV-1  PMID:15596521   
...            ...             ...            ...   
49037       Gluten           Wheat  PMID:33927715   
49038       Gluten           Wheat  PMID:33927715   
49039       Gluten           Wheat  PMID:33927715   
49040       Gluten           Wheat  PMID:33927715   
49041       Gluten           Wheat  PMID:33927715   

                                                  Method  \
0      {"frequency": "", "identification": "tetramer-...   
1      {"frequency": "", "identification": "tetramer-...   
2      {"frequency": "", "identification": "tetramer-...   
3      {"frequency": "", "identification": "tetramer-...   
4      {"frequency": "", "identification": "tetramer-...   
...                                                  ...   
49037  {"frequency": "", "identification": "tetramer-...   
49038  {"frequency": "", "identification": "tetramer-...   
49039  {"frequency": "", "identification": "tetramer-...   
49040  {"frequency": "", "identification": "tetramer-...   
49041  {"frequency": "", "identification": "tetramer-...   

                                                    Meta  \
0      {"cell.subset": "CD8+", "clone.id": "", "donor...   
1      {"cell.subset": "CD8+", "clone.id": "", "donor...   
2      {"cell.subset": "CD8+", "clone.id": "", "donor...   
3      {"cell.subset": "CD8+", "clone.id": "", "donor...   
4      {"cell.subset": "CD8+", "clone.id": "", "donor...   
...                                                  ...   
49037  {"cell.subset": "CD4+", "clone.id": "", "donor...   
49038  {"cell.subset": "CD4+", "clone.id": "", "donor...   
49039  {"cell.subset": "CD4+", "clone.id": "", "donor...   
49040  {"cell.subset": "CD4+", "clone.id": "", "donor...   
49041  {"cell.subset": "CD4+", "clone.id": "", "donor...   

                                                 CDR3fix Score      TRB_leader  
0      {"cdr3": "CASSYLPGQGDHYSNQPQHF", "cdr3_old": "...

In [21]:
vdjdb_beta_only_df.columns

Index(['TCR_name', 'Gene', 'TRB_CDR3', 'TRBV', 'TRBJ', 'Species', 'MHC A',
       'MHC B', 'MHC class', 'Epitope', 'Epitope gene', 'Epitope species',
       'Reference', 'Method', 'Meta', 'CDR3fix', 'Score', 'TRB_leader'],
      dtype='object')

In [22]:
template_stitchr.columns

Index(['TCR_name', 'TRAV', 'TRAJ', 'TRA_CDR3', 'TRBV', 'TRBJ', 'TRB_CDR3',
       'TRAC', 'TRBC', 'TRA_leader', 'TRB_leader', 'Linker', 'Link_order',
       'TRA_5_prime_seq', 'TRA_3_prime_seq', 'TRB_5_prime_seq',
       'TRB_3_prime_seq'],
      dtype='object')

In [23]:
# Assuming your DataFrame is named vdjdb_beta_only_df
# Define the new column order
new_columns = [
    'TCR_name', 'TRAV', 'TRAJ', 'TRA_CDR3', 'TRBV', 'TRBJ', 'TRB_CDR3',
    'TRAC', 'TRBC', 'TRA_leader', 'TRB_leader', 'Linker', 'Link_order',
    'TRA_5_prime_seq', 'TRA_3_prime_seq', 'TRB_5_prime_seq', 'TRB_3_prime_seq'
]

# Map your original columns to the new template, when applicable
column_mapping = {
    "TCR_name": "TCR_name",
    'TRB_CDR3': 'TRB_CDR3', # Example: Map 'TRB_CDR3' from the original DataFrame to 'TRB_CDR3' in the new DataFrame
    "TRBV": "TRBV", 
    "TRBJ": "TRBJ", 
    "TRB_leader": "TRB_leader"
}

# Use .rename() to change column names based on the mapping above
# This step is optional and depends on whether you need to rename any column to match the new template
vdjdb_beta_only_df = vdjdb_beta_only_df.rename(columns=column_mapping)

# Discard any columns not in the new column list and add missing ones with default values
vdjdb_beta_only_df = vdjdb_beta_only_df.reindex(columns=new_columns, fill_value=np.nan)

# At this point, vdjdb_beta_only_df_rearranged contains your DataFrame rearranged with the new column order,
# discarding columns not in the template, and adding new columns with default values np.nan where necessary.
vdjdb_beta_only_df.columns

Index(['TCR_name', 'TRAV', 'TRAJ', 'TRA_CDR3', 'TRBV', 'TRBJ', 'TRB_CDR3',
       'TRAC', 'TRBC', 'TRA_leader', 'TRB_leader', 'Linker', 'Link_order',
       'TRA_5_prime_seq', 'TRA_3_prime_seq', 'TRB_5_prime_seq',
       'TRB_3_prime_seq'],
      dtype='object')

In [24]:
vdjdb_beta_only_df.to_csv("../data/VDJdb/temp/vdjdb_beta_only_Stitchr_schema.tsv")

In [25]:
from_beta_vdjdb_path = "../data/VDJdb/temp/vdjdb_beta_only_Stitchr_schema.tsv"
to_beta_vdjdb_path =  "../data/StitchrOutputs/VDJdb_Stitchr/VDJdb_beta_Stitchr"
command = ['thimble', '-in', from_beta_vdjdb_path, '-r', 'B', "-s", "HUMAN", "-o", to_beta_vdjdb_path]

In [26]:
result = subprocess.run(command, capture_output=True, text=True)

# Check if the command was executed successfully
if result.returncode == 0:
    print("Command executed successfully.")
    # If you want to print the output
    print("Output:", result.stdout)
else:
    print("Error in command execution.")
    # If you want to print the error
    print("Error:", result.stderr)

Error in command execution.
Error: ValueError: Headers in input file don't match the expectations - please check template.



-------------------------------------------------------------------------------------------------
-------------------------------------------------------------------------------------------------

In the follwoing we want to create the same output for the McPAS-TCR dataset!

In [27]:
mcpastcr_artifact = wandb.use_artifact(f"{MCPAS_TCR}:latest")
mcpastcr_table = mcpastcr_artifact.get(f"{MCPAS_TCR}_table.table.json")
mcpastcr_artifact_dir = mcpastcr_artifact.download()
mcpastcr_df = pd.DataFrame(data=mcpastcr_table.data, columns=mcpastcr_table.columns)
mcpastcr_df

wandb:   1 of 1 files downloaded.  
wandb:   1 of 1 files downloaded.  


CDR3.alpha.aa      CDR3.beta.aa Species   Category  \
0                 nan     CASSDAGANTEVF   Mouse  Pathogens   
1                 nan     CASSDAGAYAEQF   Mouse  Pathogens   
2                 nan     CASSDAGGAAEVF   Mouse  Pathogens   
3                 nan     CASSDAGHSPLYF   Mouse  Pathogens   
4                 nan    CASSDAWGGAEQYF   Mouse  Pathogens   
...               ...               ...     ...        ...   
39980   CAGSGAGSYQLTF  CASSLEGQASSYEQYF   Human  Pathogens   
39981   CAGLGAGSYQLTF  CASSLEGQGASYEQYF   Human  Pathogens   
39982  CATEGNSGYSTLTF    CASSSQGGNYGYTF   Human  Pathogens   
39983  CATEGDSGYSTLTF    CASSYQGGNYGYTF   Human  Pathogens   
39984  CASEGNSGYSTLTF    CASSFQGGNYGYTF   Human  Pathogens   

                                       Pathology Pathology.Mesh.ID  \
0      Lymphocytic choriomeningitis virus (LCMV)           D008217   
1      Lymphocytic choriomeningitis virus (LCMV)           D008217   
2      Lymphocytic choriomeningitis virus (LCMV)           D008217   
3      Lymphocytic choriomeningitis virus (LCMV)           D008217   
4      Lymphocytic choriomeningitis virus (LCMV)           D008217   
...                                          ...               ...   
39980                   Epstein Barr virus (EBV)           D020031   
39981                   Epstein Barr virus (EBV)           D020031   
39982                   Epstein Barr virus (EBV)           D020031   
39983                   Epstein Barr virus (EBV)           D020031   
39984                   Epstein Barr virus (EBV)           D020031   

      Additional.study.details Antigen.identification.method Single.cell  NGS  \
0                          nan                           2.1          No   No   
1                          nan                           2.1          No   No   
2                          nan                           2.1          No   No   
3                          nan                           2.1          No   No   
4                          nan                           2.1          No   No   
...                        ...                           ...         ...  ...   
39980                      nan                           1.0         Yes  Yes   
39981                      nan                           1.0         Yes  Yes   
39982                      nan                           1.0         Yes  Yes   
39983                      nan                           1.0         Yes  Yes   
39984                      nan                           1.0         Yes  Yes   

       ...       TRAV       TRAJ     TRBV TRBD     TRBJ  \
0      ...        nan        nan  TRBV8-1  nan  TRBJ1-1   
1      ...        nan        nan  TRBV8-1  nan  TRBJ2-1   
2      ...        nan        nan  TRBV8-3  nan  TRBJ1-1   
3      ...        nan        nan  TRBV8-1  nan  TRBJ1-6   
4      ...        nan        nan  TRBV8-3  nan  TRBJ2-6   
...    ...        ...        ...      ...  ...      ...   
39980  ...     TRAV25     TRAJ28      nan  nan      nan   
39981  ...     TRAV25     TRAJ28      nan  nan      nan   
39982  ...  TRAV17*01  TRAJ11*01      nan  nan      nan   
39983  ...  TRAV17*01  TRAJ11*01      nan  nan      nan   
39984  ...  TRAV17*01  TRAJ11*01      nan  nan      nan   

      Reconstructed.J.annotation CDR3.beta.nt Mouse.strain PubMed.ID Remarks  
0                             No          nan          P14   1716213     nan  
1                             No          nan          P14   1716213     nan  
2                             No          nan          P14   1716213     nan  
3                             No          nan          P14   1716213     nan  
4                             No          nan          P14   1716213     nan  
...                          ...          ...          ...       ...     ...  
39980                        nan          nan          nan  35401538     nan  
39981                        nan          nan          nan  35401538     nan  
39982                        

In [28]:
mcpastcr_df.columns

Index(['CDR3.alpha.aa', 'CDR3.beta.aa', 'Species', 'Category', 'Pathology',
       'Pathology.Mesh.ID', 'Additional.study.details',
       'Antigen.identification.method', 'Single.cell', 'NGS',
       'Antigen.protein', 'Protein.ID', 'Epitope.peptide', 'Epitope.ID', 'MHC',
       'Tissue', 'T.Cell.Type', 'T.cell.characteristics', 'CDR3.alpha.nt',
       'TRAV', 'TRAJ', 'TRBV', 'TRBD', 'TRBJ', 'Reconstructed.J.annotation',
       'CDR3.beta.nt', 'Mouse.strain', 'PubMed.ID', 'Remarks'],
      dtype='object')

In [29]:
template_stitchr.columns

Index(['TCR_name', 'TRAV', 'TRAJ', 'TRA_CDR3', 'TRBV', 'TRBJ', 'TRB_CDR3',
       'TRAC', 'TRBC', 'TRA_leader', 'TRB_leader', 'Linker', 'Link_order',
       'TRA_5_prime_seq', 'TRA_3_prime_seq', 'TRB_5_prime_seq',
       'TRB_3_prime_seq'],
      dtype='object')

In [30]:
rename_mapping = {
    'CDR3.alpha.aa': 'TRA_CDR3',
    'CDR3.beta.aa': 'TRB_CDR3',
}
mcpastcr_df_renamed = mcpastcr_df.rename(columns=rename_mapping)
mcpastcr_df_renamed.columns

Index(['TRA_CDR3', 'TRB_CDR3', 'Species', 'Category', 'Pathology',
       'Pathology.Mesh.ID', 'Additional.study.details',
       'Antigen.identification.method', 'Single.cell', 'NGS',
       'Antigen.protein', 'Protein.ID', 'Epitope.peptide', 'Epitope.ID', 'MHC',
       'Tissue', 'T.Cell.Type', 'T.cell.characteristics', 'CDR3.alpha.nt',
       'TRAV', 'TRAJ', 'TRBV', 'TRBD', 'TRBJ', 'Reconstructed.J.annotation',
       'CDR3.beta.nt', 'Mouse.strain', 'PubMed.ID', 'Remarks'],
      dtype='object')

In [31]:
mcpastcr_df_renamed

TRA_CDR3          TRB_CDR3 Species   Category  \
0                 nan     CASSDAGANTEVF   Mouse  Pathogens   
1                 nan     CASSDAGAYAEQF   Mouse  Pathogens   
2                 nan     CASSDAGGAAEVF   Mouse  Pathogens   
3                 nan     CASSDAGHSPLYF   Mouse  Pathogens   
4                 nan    CASSDAWGGAEQYF   Mouse  Pathogens   
...               ...               ...     ...        ...   
39980   CAGSGAGSYQLTF  CASSLEGQASSYEQYF   Human  Pathogens   
39981   CAGLGAGSYQLTF  CASSLEGQGASYEQYF   Human  Pathogens   
39982  CATEGNSGYSTLTF    CASSSQGGNYGYTF   Human  Pathogens   
39983  CATEGDSGYSTLTF    CASSYQGGNYGYTF   Human  Pathogens   
39984  CASEGNSGYSTLTF    CASSFQGGNYGYTF   Human  Pathogens   

                                       Pathology Pathology.Mesh.ID  \
0      Lymphocytic choriomeningitis virus (LCMV)           D008217   
1      Lymphocytic choriomeningitis virus (LCMV)           D008217   
2      Lymphocytic choriomeningitis virus (LCMV)           D008217   
3      Lymphocytic choriomeningitis virus (LCMV)           D008217   
4      Lymphocytic choriomeningitis virus (LCMV)           D008217   
...                                          ...               ...   
39980                   Epstein Barr virus (EBV)           D020031   
39981                   Epstein Barr virus (EBV)           D020031   
39982                   Epstein Barr virus (EBV)           D020031   
39983                   Epstein Barr virus (EBV)           D020031   
39984                   Epstein Barr virus (EBV)           D020031   

      Additional.study.details Antigen.identification.method Single.cell  NGS  \
0                          nan                           2.1          No   No   
1                          nan                           2.1          No   No   
2                          nan                           2.1          No   No   
3                          nan                           2.1          No   No   
4                          nan                           2.1          No   No   
...                        ...                           ...         ...  ...   
39980                      nan                           1.0         Yes  Yes   
39981                      nan                           1.0         Yes  Yes   
39982                      nan                           1.0         Yes  Yes   
39983                      nan                           1.0         Yes  Yes   
39984                      nan                           1.0         Yes  Yes   

       ...       TRAV       TRAJ     TRBV TRBD     TRBJ  \
0      ...        nan        nan  TRBV8-1  nan  TRBJ1-1   
1      ...        nan        nan  TRBV8-1  nan  TRBJ2-1   
2      ...        nan        nan  TRBV8-3  nan  TRBJ1-1   
3      ...        nan        nan  TRBV8-1  nan  TRBJ1-6   
4      ...        nan        nan  TRBV8-3  nan  TRBJ2-6   
...    ...        ...        ...      ...  ...      ...   
39980  ...     TRAV25     TRAJ28      nan  nan      nan   
39981  ...     TRAV25     TRAJ28      nan  nan      nan   
39982  ...  TRAV17*01  TRAJ11*01      nan  nan      nan   
39983  ...  TRAV17*01  TRAJ11*01      nan  nan      nan   
39984  ...  TRAV17*01  TRAJ11*01      nan  nan      nan   

      Reconstructed.J.annotation CDR3.beta.nt Mouse.strain PubMed.ID Remarks  
0                             No          nan          P14   1716213     nan  
1                             No          nan          P14   1716213     nan  
2                             No          nan          P14   1716213     nan  
3                             No          nan          P14   1716213     nan  
4                             No          nan          P14   1716213     nan  
...                          ...          ...          ...       ...     ...  
39980                        nan          nan          nan  35401538     nan  
39981                        nan          nan          nan  35401538     nan  
39982                        nan  

In [32]:
mcpastcr_only_human = mcpastcr_df_renamed[mcpastcr_df_renamed["Species"] == "Human"]
mcpastcr_only_human

TRA_CDR3            TRB_CDR3 Species   Category  \
74       CALIPGGQKLLF     CAAGETSGVSYNEQF   Human     Cancer   
75       CAENMMNTGNQF  CASRPTITVPYSNQPQHF   Human     Cancer   
76          CAYRGLGVV  CASRPTITVPYSNQPQHF   Human     Cancer   
77     CAASKGGSQGNLIF  CASSLVVWDRGGNOPQHF   Human     Cancer   
78         CAVGATGNQF      CASSQDLLSWDEQF   Human     Cancer   
...               ...                 ...     ...        ...   
39980   CAGSGAGSYQLTF    CASSLEGQASSYEQYF   Human  Pathogens   
39981   CAGLGAGSYQLTF    CASSLEGQGASYEQYF   Human  Pathogens   
39982  CATEGNSGYSTLTF      CASSSQGGNYGYTF   Human  Pathogens   
39983  CATEGDSGYSTLTF      CASSYQGGNYGYTF   Human  Pathogens   
39984  CASEGNSGYSTLTF      CASSFQGGNYGYTF   Human  Pathogens   

                      Pathology Pathology.Mesh.ID Additional.study.details  \
74                     Melanoma           D008545                      nan   
75                     Melanoma           D008545                      nan   
76                     Melanoma           D008545                      nan   
77                     Melanoma           D008545                      nan   
78                     Melanoma           D008545                      nan   
...                         ...               ...                      ...   
39980  Epstein Barr virus (EBV)           D020031                      nan   
39981  Epstein Barr virus (EBV)           D020031                      nan   
39982  Epstein Barr virus (EBV)           D020031                      nan   
39983  Epstein Barr virus (EBV)           D020031                      nan   
39984  Epstein Barr virus (EBV)           D020031                      nan   

      Antigen.identification.method Single.cell  NGS  ...       TRAV  \
74                              2.2          No   No  ...      TRAV9   
75                              2.2          No   No  ...    TRAV8-2   
76                              2.2          No   No  ...   TRAV14-1   
77                              2.2          No   No  ...     TRAV17   
78                              2.2          No   No  ...      TRAV1   
...                             ...         ...  ...  ...        ...   
39980                           1.0         Yes  Yes  ...     TRAV25   
39981                           1.0         Yes  Yes  ...     TRAV25   
39982                           1.0         Yes  Yes  ...  TRAV17*01   
39983                           1.0         Yes  Yes  ...  TRAV17*01   
39984                           1.0         Yes  Yes  ...  TRAV17*01   

            TRAJ    TRBV     TRBD     TRBJ Reconstructed.J.annotation  \
74        TRAJ16  TRBV22  TRBD2-1  TRBJ2-1                         No   
75        TRAJ49  TRBV13  TRBD1-1  TRBJ1-5                         No   
76        TRAJ32  TRBV13  TRBD1-1  TRBJ1-5                         No   
77        TRAJ42   TRBV6  TRBD1-1  TRBJ1-5                         No   
78        TRAJ49   TRBV7  TRBD2-1  TRBJ2-1                         No   
...          ...     ...      ...      ...                        ...   
39980     TRAJ28     nan      nan      nan                        nan   
39981     TRAJ28     nan      nan      nan                        nan   
39982  TRAJ11*01     nan      nan      nan                        nan   
39983  TRAJ11*01     nan      nan      nan                        nan   
39984  TRAJ11*01     nan      nan      nan                        nan   

                                            CDR3.beta.nt Mouse.strain  \
74          GTGCCGCTGGGGAGACTAGCGGGGTGTCGTACAATGAGCAGTTC          nan   
75     GTGCCTGTGCCAGCCGACCTACTATAACGGTCCCGTATAGCAATCA...          nan   
76     GTGCCTGTGCCAGCCGACCTACTATAACGGTCCCGTATAGCAATCA...          nan   
77     GTGCCAGCAGCTTAGTAGTCTGGGACAGGGGTGGTAATCAGCCCCA...          nan   
78           TGTGCCAGCAGCTCAAGATCTCCTGAGTTGGGATGAGCAGTTC          nan   
...                                                  ...          ...   
39980                                                n

In [33]:
# Define the new column order and include new columns
new_columns = [
    'TCR_name', 'TRAV', 'TRAJ', 'TRA_CDR3', 'TRBV', 'TRBJ', 'TRB_CDR3',
    'TRAC', 'TRBC', 'TRA_leader', 'TRB_leader', 'Linker', 'Link_order',
    'TRA_5_prime_seq', 'TRA_3_prime_seq', 'TRB_5_prime_seq', 'TRB_3_prime_seq'
]

# Map existing columns to the new schema, filling non-existing columns with NaN
# Note: This template uses placeholders for TRA_CDR3 and TRB_CDR3, which you might need to adjust
# based on your actual column names or desired mappings.
mcpastcr_df_template = mcpastcr_df_renamed.reindex(columns=new_columns).assign(
    TRAC=np.nan,
    TRBC=np.nan,
    Linker=np.nan,
    Link_order=np.nan,
    TRA_5_prime_seq=np.nan,
    TRA_3_prime_seq=np.nan,
    TRB_5_prime_seq=np.nan,
    TRB_3_prime_seq=np.nan
)

# For the TRA_leader and TRB_leader, we add "(L)" to the TRAV and TRBV values, respectively
# Adjust these lines if TRAV and TRBV are not the exact column names or if additional logic is needed
mcpastcr_df_template['TRA_leader'] = mcpastcr_df['TRAV'] + "(L)"
mcpastcr_df_template['TRB_leader'] = mcpastcr_df['TRBV'] + "(L)"

# here replace the string 'nan' with np.nan as np.nan should be used with pandas dataframes
mcpastcr_df_template = mcpastcr_df_template.replace('nan', np.nan) 
mcpastcr_df_template['TCR_name'] = range(1, len(mcpastcr_df) + 1) # use index as TCR_name
mcpastcr_df_template

TCR_name       TRAV       TRAJ        TRA_CDR3     TRBV     TRBJ  \
0             1        NaN        NaN             NaN  TRBV8-1  TRBJ1-1   
1             2        NaN        NaN             NaN  TRBV8-1  TRBJ2-1   
2             3        NaN        NaN             NaN  TRBV8-3  TRBJ1-1   
3             4        NaN        NaN             NaN  TRBV8-1  TRBJ1-6   
4             5        NaN        NaN             NaN  TRBV8-3  TRBJ2-6   
...         ...        ...        ...             ...      ...      ...   
39980     39981     TRAV25     TRAJ28   CAGSGAGSYQLTF      NaN      NaN   
39981     39982     TRAV25     TRAJ28   CAGLGAGSYQLTF      NaN      NaN   
39982     39983  TRAV17*01  TRAJ11*01  CATEGNSGYSTLTF      NaN      NaN   
39983     39984  TRAV17*01  TRAJ11*01  CATEGDSGYSTLTF      NaN      NaN   
39984     39985  TRAV17*01  TRAJ11*01  CASEGNSGYSTLTF      NaN      NaN   

               TRB_CDR3  TRAC  TRBC    TRA_leader  TRB_leader  Linker  \
0         CASSDAGANTEVF   NaN   NaN        nan(L)  TRBV8-1(L)     NaN   
1         CASSDAGAYAEQF   NaN   NaN        nan(L)  TRBV8-1(L)     NaN   
2         CASSDAGGAAEVF   NaN   NaN        nan(L)  TRBV8-3(L)     NaN   
3         CASSDAGHSPLYF   NaN   NaN        nan(L)  TRBV8-1(L)     NaN   
4        CASSDAWGGAEQYF   NaN   NaN        nan(L)  TRBV8-3(L)     NaN   
...                 ...   ...   ...           ...         ...     ...   
39980  CASSLEGQASSYEQYF   NaN   NaN     TRAV25(L)      nan(L)     NaN   
39981  CASSLEGQGASYEQYF   NaN   NaN     TRAV25(L)      nan(L)     NaN   
39982    CASSSQGGNYGYTF   NaN   NaN  TRAV17*01(L)      nan(L)     NaN   
39983    CASSYQGGNYGYTF   NaN   NaN  TRAV17*01(L)      nan(L)     NaN   
39984    CASSFQGGNYGYTF   NaN   NaN  TRAV17*01(L)      nan(L)     NaN   

       Link_order  TRA_5_prime_seq  TRA_3_prime_seq  TRB_5_prime_seq  \
0             NaN              NaN              NaN              NaN   
1             NaN              NaN              NaN              NaN   
2             NaN              NaN              NaN              NaN   
3             NaN              NaN              NaN              NaN   
4             NaN              NaN              NaN              NaN   
...           ...              ...              ...              ...   
39980         NaN              NaN              NaN              NaN   
39981         NaN              NaN              NaN              NaN   
39982         NaN              NaN              NaN              NaN   
39983         NaN              NaN              NaN              NaN   
39984         NaN              NaN              NaN              NaN   

       TRB_3_prime_seq  
0                  NaN  
1                  NaN  
2                  NaN  
3                  NaN  
4                  NaN  
...                ...  
39980              NaN  
39981              NaN  
39982              NaN  
39983              NaN  
39984              NaN  

[39985 rows x 17 columns]

In [34]:
# mcpastcr_df_template.to_csv("./McPAS-TCR_complete_onlyHumans.tsv", sep="\t", index=False)
mcpastcr_df_template.columns

Index(['TCR_name', 'TRAV', 'TRAJ', 'TRA_CDR3', 'TRBV', 'TRBJ', 'TRB_CDR3',
       'TRAC', 'TRBC', 'TRA_leader', 'TRB_leader', 'Linker', 'Link_order',
       'TRA_5_prime_seq', 'TRA_3_prime_seq', 'TRB_5_prime_seq',
       'TRB_3_prime_seq'],
      dtype='object')

In [35]:
# Define columns related to alpha and beta chains for filtering
alpha_columns = ['TRA_CDR3']
beta_columns = ['TRB_CDR3']

# Check for presence of non-NaN values in alpha and beta columns
alpha_present = mcpastcr_df_template[alpha_columns].notna().any(axis=1)
beta_present = mcpastcr_df_template[beta_columns].notna().any(axis=1)

alpha_present

0        False
1        False
2        False
3        False
4        False
         ...  
39980     True
39981     True
39982     True
39983     True
39984     True
Length: 39985, dtype: bool

In [36]:
beta_present

0        True
1        True
2        True
3        True
4        True
         ... 
39980    True
39981    True
39982    True
39983    True
39984    True
Length: 39985, dtype: bool

The output above indicates that McPAS-TCR only contains paired or beta-only information!

In [37]:
alpha_only_df = mcpastcr_df_template[alpha_present & ~beta_present]
beta_only_df = mcpastcr_df_template[~alpha_present & beta_present]
paired_df = mcpastcr_df_template[alpha_present & beta_present]

In [38]:
beta_only_df

TCR_name TRAV TRAJ TRA_CDR3      TRBV     TRBJ        TRB_CDR3  TRAC  \
0             1  NaN  NaN      NaN   TRBV8-1  TRBJ1-1   CASSDAGANTEVF   NaN   
1             2  NaN  NaN      NaN   TRBV8-1  TRBJ2-1   CASSDAGAYAEQF   NaN   
2             3  NaN  NaN      NaN   TRBV8-3  TRBJ1-1   CASSDAGGAAEVF   NaN   
3             4  NaN  NaN      NaN   TRBV8-1  TRBJ1-6   CASSDAGHSPLYF   NaN   
4             5  NaN  NaN      NaN   TRBV8-3  TRBJ2-6  CASSDAWGGAEQYF   NaN   
...         ...  ...  ...      ...       ...      ...             ...   ...   
39964     39965  NaN  NaN      NaN   TRBV6-6  TRBJ1-6  CASSYPGNNSPLHF   NaN   
39965     39966  NaN  NaN      NaN  TRBV20-1  TRBJ2-7    CSARASVATSST   NaN   
39966     39967  NaN  NaN      NaN    TRBV19  TRBJ2-5   CATSSAAQETQYF   NaN   
39967     39968  NaN  NaN      NaN  TRBV10-3  TRBJ1-2   CAINPGTAYGYTF   NaN   
39968     39969  NaN  NaN      NaN    TRBV18  TRBJ2-1    CASSPGTPEQFF   NaN   

       TRBC TRA_leader   TRB_leader  Linker  Link_order  TRA_5_prime_seq  \
0       NaN     nan(L)   TRBV8-1(L)     NaN         NaN              NaN   
1       NaN     nan(L)   TRBV8-1(L)     NaN         NaN              NaN   
2       NaN     nan(L)   TRBV8-3(L)     NaN         NaN              NaN   
3       NaN     nan(L)   TRBV8-1(L)     NaN         NaN              NaN   
4       NaN     nan(L)   TRBV8-3(L)     NaN         NaN              NaN   
...     ...        ...          ...     ...         ...              ...   
39964   NaN     nan(L)   TRBV6-6(L)     NaN         NaN              NaN   
39965   NaN     nan(L)  TRBV20-1(L)     NaN         NaN              NaN   
39966   NaN     nan(L)    TRBV19(L)     NaN         NaN              NaN   
39967   NaN     nan(L)  TRBV10-3(L)     NaN         NaN              NaN   
39968   NaN     nan(L)    TRBV18(L)     NaN         NaN              NaN   

       TRA_3_prime_seq  TRB_5_prime_seq  TRB_3_prime_seq  
0                  NaN              NaN              NaN  
1                  NaN              NaN              NaN  
2                  NaN              NaN              NaN  
3                  NaN              NaN              NaN  
4                  NaN              NaN              NaN  
...                ...              ...              ...  
39964              NaN              NaN              NaN  
39965              NaN              NaN              NaN  
39966              NaN              NaN              NaN  
39967              NaN              NaN              NaN  
39968              NaN              NaN              NaN  

[24875 rows x 17 columns]

In [39]:
beta_only_df.to_csv("../data/McPAS-TCR/temp/McPAS-TCR_beta_only_Stitchr_schema.tsv", sep="\t", index=False)

**THIS IS OFF!!! => TRBV and TRBJ MUST HAVE SOME VALID VALUES!!**

=> after considerations i think now that it is possible that some V and/or J values are NaN even if the CDR3 region is not NaN. In VDJdb some have this as well. 

In [40]:
paired_df

TCR_name       TRAV       TRAJ        TRA_CDR3      TRBV     TRBJ  \
14           15        NaN        NaN    CILRAGYQNFYF   TRBV8-2  TRBJ1-5   
15           16        NaN        NaN    CILRAGYQNFYF  TRBV13-2  TRBJ1-5   
16           17        NaN        NaN  CALSDGEHGLQTYF   TRBV8-2  TRBJ2-2   
17           18        NaN        NaN  CALSDGEHGLQTYF  TRBV13-2  TRBJ2-2   
19           20        NaN        NaN  CALSDQGGADRLTF     TRBV6  TRBJ2-2   
...         ...        ...        ...             ...       ...      ...   
39980     39981     TRAV25     TRAJ28   CAGSGAGSYQLTF       NaN      NaN   
39981     39982     TRAV25     TRAJ28   CAGLGAGSYQLTF       NaN      NaN   
39982     39983  TRAV17*01  TRAJ11*01  CATEGNSGYSTLTF       NaN      NaN   
39983     39984  TRAV17*01  TRAJ11*01  CATEGDSGYSTLTF       NaN      NaN   
39984     39985  TRAV17*01  TRAJ11*01  CASEGNSGYSTLTF       NaN      NaN   

               TRB_CDR3  TRAC  TRBC    TRA_leader   TRB_leader  Linker  \
14         CASGDGNQAPLF   NaN   NaN        nan(L)   TRBV8-2(L)     NaN   
15         CASGDGNQAPLF   NaN   NaN        nan(L)  TRBV13-2(L)     NaN   
16     CASGGTGGANTGQLYF   NaN   NaN        nan(L)   TRBV8-2(L)     NaN   
17     CASGGTGGANTGQLYF   NaN   NaN        nan(L)  TRBV13-2(L)     NaN   
19       CASIPTANTGQLYF   NaN   NaN        nan(L)     TRBV6(L)     NaN   
...                 ...   ...   ...           ...          ...     ...   
39980  CASSLEGQASSYEQYF   NaN   NaN     TRAV25(L)       nan(L)     NaN   
39981  CASSLEGQGASYEQYF   NaN   NaN     TRAV25(L)       nan(L)     NaN   
39982    CASSSQGGNYGYTF   NaN   NaN  TRAV17*01(L)       nan(L)     NaN   
39983    CASSYQGGNYGYTF   NaN   NaN  TRAV17*01(L)       nan(L)     NaN   
39984    CASSFQGGNYGYTF   NaN   NaN  TRAV17*01(L)       nan(L)     NaN   

       Link_order  TRA_5_prime_seq  TRA_3_prime_seq  TRB_5_prime_seq  \
14            NaN              NaN              NaN              NaN   
15            NaN              NaN              NaN              NaN   
16            NaN              NaN              NaN              NaN   
17            NaN              NaN              NaN              NaN   
19            NaN              NaN              NaN              NaN   
...           ...              ...              ...              ...   
39980         NaN              NaN              NaN              NaN   
39981         NaN              NaN              NaN              NaN   
39982         NaN              NaN              NaN              NaN   
39983         NaN              NaN              NaN              NaN   
39984         NaN              NaN              NaN              NaN   

       TRB_3_prime_seq  
14                 NaN  
15                 NaN  
16                 NaN  
17                 NaN  
19                 NaN  
...                ...  
39980              NaN  
39981              NaN  
39982              NaN  
39983              NaN  
39984              NaN  

[12739 rows x 17 columns]

In [41]:
paired_df.to_csv("../data/McPAS-TCR/temp/McPAS-TCR_paired_only_Stitchr_schema.tsv", sep="\t", index=False)

What confuses me a bit is this: 
paired + beta-only = 12739 + 24875 = 37'614

Total McPAS-TCR = 39985 
=> where are the "missing" rows? What do they contain that they nowhere got included? 

In [42]:
alpha_only_df

TCR_name      TRAV    TRAJ           TRA_CDR3 TRBV TRBJ TRB_CDR3  TRAC  \
101         102     TRAV8  TRAJ48            CADWGQF  NaN  NaN      NaN   NaN   
102         103     TRAV2   TRAJ1         CAVKDSTLTF  NaN  NaN      NaN   NaN   
268         269       NaN  TRAJ32       CALAFNGAKLTF  NaN  NaN      NaN   NaN   
269         270     TRAV4  TRAJ30      CALSEGNTGKLIF  NaN  NaN      NaN   NaN   
270         271    TRAV11  TRAJ27      CAAEASNTNKVVF  NaN  NaN      NaN   NaN   
...         ...       ...     ...                ...  ...  ...      ...   ...   
39514     39515  TRAV26-1  TRAJ20        CIAPFGNDMRF  NaN  NaN      NaN   NaN   
39515     39516  TRAV26-1  TRAJ43  CLVGDLDGSSFKGKLIF  NaN  NaN      NaN   NaN   
39516     39517     TRAV4  TRAJ37    CAMRVSGGSNYKLTF  NaN  NaN      NaN   NaN   
39518     39519       NaN     NaN     CLVGDDTGFQKLVF  NaN  NaN      NaN   NaN   
39519     39520     TRAV4   TRAJ8  CAGPSHAGGTSYGKLTF  NaN  NaN      NaN   NaN   

       TRBC   TRA_leader TRB_leader  Linker  Link_order  TRA_5_prime_seq  \
101     NaN     TRAV8(L)     nan(L)     NaN         NaN              NaN   
102     NaN     TRAV2(L)     nan(L)     NaN         NaN              NaN   
268     NaN       nan(L)     nan(L)     NaN         NaN              NaN   
269     NaN     TRAV4(L)     nan(L)     NaN         NaN              NaN   
270     NaN    TRAV11(L)     nan(L)     NaN         NaN              NaN   
...     ...          ...        ...     ...         ...              ...   
39514   NaN  TRAV26-1(L)     nan(L)     NaN         NaN              NaN   
39515   NaN  TRAV26-1(L)     nan(L)     NaN         NaN              NaN   
39516   NaN     TRAV4(L)     nan(L)     NaN         NaN              NaN   
39518   NaN       nan(L)     nan(L)     NaN         NaN              NaN   
39519   NaN     TRAV4(L)     nan(L)     NaN         NaN              NaN   

       TRA_3_prime_seq  TRB_5_prime_seq  TRB_3_prime_seq  
101                NaN              NaN              NaN  
102                NaN              NaN              NaN  
268                NaN              NaN              NaN  
269                NaN              NaN              NaN  
270                NaN              NaN              NaN  
...                ...              ...              ...  
39514              NaN              NaN              NaN  
39515              NaN              NaN              NaN  
39516              NaN              NaN              NaN  
39518              NaN              NaN              NaN  
39519              NaN              NaN              NaN  

[1970 rows x 17 columns]

even if they would be alpha-only we do not match the toal number... 

paired + beta-only + alpha-only = 12739 + 24875 + 1970 = 39'584

-----------------------------------------------------------------------------------------------------------

In the following we pass the McPAS-TCR data through the Stitchr library!

In [43]:
from_paired_mcpas_tcr_path = "../data/McPAS-TCR/temp/McPAS-TCR_paired_only_Stitchr_schema.tsv"
to_paired_mcpas_tcr_path =  "../data/StitchrOutputs/McPAS-TCR_Stitchr/McPAS-TCR_paired_Stitchr"
command = ['thimble', '-in', from_paired_mcpas_tcr_path, '-r', 'AB', "-s", "HUMAN", "-o", to_paired_mcpas_tcr_path]

In [44]:
result = subprocess.run(command, capture_output=True, text=True)

# Check if the command was executed successfully
if result.returncode == 0:
    print("Command executed successfully.")
    # If you want to print the output
    print("Output:", result.stdout)
else:
    print("Error in command execution.")
    # If you want to print the error
    print("Error:", result.stderr)

Command executed successfully.
Output: Took 0.37 seconds



In [45]:
from_beta_mcpas_tcr_path = "../data/McPAS-TCR/temp/McPAS-TCR_beta_only_Stitchr_schema.tsv"
to_beta_mcpas_tcr_path =  "../data/StitchrOutputs/McPAS-TCR_Stitchr/McPAS-TCR_paired_Stitchr"
command = ['thimble', '-in', from_beta_mcpas_tcr_path, '-r', 'B', "-s", "HUMAN", "-o", to_beta_mcpas_tcr_path]

In [46]:
result = subprocess.run(command, capture_output=True, text=True)

# Check if the command was executed successfully
if result.returncode == 0:
    print("Command executed successfully.")
    # If you want to print the output
    print("Output:", result.stdout)
else:
    print("Error in command execution.")
    # If you want to print the error
    print("Error:", result.stderr)

Command executed successfully.
Output: Took 0.66 seconds

